source/adios2/helper/adiosMpiHandshake.cpp

```diff
@@ -0,0 +1,286 @@
+                             const int timeoutSeconds,
+                             const size_t maxStreamsPerApp,
+                             const size_t maxFilenameLength,
+                             const size_t rendezvousAppCountForStream,
+                             MPI_Comm localComm)
+{
+
+    // initialize variables
+
+    if (filename.size() > maxFilenameLength)
+    {
+        throw(std::runtime_error("Filename too long"));
+    }
+
+    MPI_Comm_size(MPI_COMM_WORLD, &m_WorldSize);
+    MPI_Comm_rank(MPI_COMM_WORLD, &m_WorldRank);
+    MPI_Comm_size(localComm, &m_LocalSize);
+    MPI_Comm_rank(localComm, &m_LocalRank);
+    m_MaxStreamsPerApp = maxStreamsPerApp;
+    m_MaxFilenameLength = maxFilenameLength;
+    m_RendezvousAppCounts[filename] = rendezvousAppCountForStream;
+
+    m_SendRequests.resize(m_WorldSize);
+    m_RecvRequests.resize(m_WorldSize);
+    for (int rank = 0; rank < m_WorldSize; ++rank)
+    {
+        m_SendRequests[rank].resize(maxStreamsPerApp);
+        m_RecvRequests[rank].resize(maxStreamsPerApp);
+    }
+
+    m_ItemSize = maxFilenameLength + sizeof(char) + sizeof(int) * 2;
+    m_Buffer.resize(m_WorldSize * maxStreamsPerApp * m_ItemSize);
+
+    // broadcast local master rank's world rank to use as app ID
+
+    if (m_LocalRank == 0)
+    {
+        m_LocalMasterRank = m_WorldRank;
+    }
+    MPI_Bcast(&m_LocalMasterRank, 1, MPI_INT, 0, localComm);
+
+    // start receiving
+
+    for (int rank = 0; rank < m_WorldSize; ++rank)
+    {
+        for (size_t stream = 0; stream < maxStreamsPerApp; ++stream)
+        {
+            MPI_Irecv(m_Buffer.data() + PlaceInBuffer(stream, rank), m_ItemSize,
+                      MPI_CHAR, rank, rank, MPI_COMM_WORLD,
+                      &m_RecvRequests[rank][stream]);
+        }
+    }
+
+    // start sending
+
+    size_t offset = 0;
+    std::vector<char> buffer(m_ItemSize);
+    std::memcpy(buffer.data(), &mode, sizeof(char));
+    offset += sizeof(char);
+    std::memcpy(buffer.data() + offset, &m_LocalMasterRank, sizeof(int));
+    offset += sizeof(int);
+    std::memcpy(buffer.data() + offset, &m_LocalSize, sizeof(int));
+    offset += sizeof(int);
+    std::memcpy(buffer.data() + offset, filename.data(), filename.size());
+
+    for (int rank = 0; rank < m_WorldSize; ++rank)
+    {
+        MPI_Isend(buffer.data(), m_ItemSize, MPI_CHAR, rank, m_WorldRank,
+                  MPI_COMM_WORLD, &m_SendRequests[rank][m_StreamID]);
+    }
+
+    // wait and check if required RendezvousAppCount reached
+
+    auto start_time = std::chrono::system_clock::now();
+    while (!Check(filename))
+    {
+        auto now_time = std::chrono::system_clock::now();
+        auto duration = std::chrono::duration_cast<std::chrono::seconds>(
+            now_time - start_time);
+        if (duration.count() > timeoutSeconds)
+        {
+            throw(std::runtime_error("Mpi handshake timeout"));
+        }
```

Comment:
There should be some sort of short sleep here to ensure the process doesn't go into a blocking spin wait.  Without the sleep then  you run the risk in an oversubscription scenario of one process completely consuming the CPU code and not allowing others to proceed.  Something like `std::this_thread::sleep_for(100ms);` should be sufficient.

Reply:
Done